examples/rl-ppo-continuous.rs

```diff
@@ -0,0 +1,272 @@
+            let log_prob = {
+                let variance = std.clone().powi(2);
+                -(action.clone() - mu).powi(2) / (variance * 2.0)
+                    - std.ln()
+                    - (2.0f32 * f32::consts::PI).sqrt().ln()
+            };
+            (log_prob /*(std, mu, action)*/, action)
+        };
+
+        let (new_state, reward, done) = step_simulation(&dev, state.clone(), &action);
+
+        // <calc advantage>
+        let td_target: Tensor1D<ACTION_SIZE> =
+            net.value.forward(net.l1.forward(new_state.clone())) * f32::from(done as u8) * GAMMA
+                + reward;
+        let delta: Tensor1D<ACTION_SIZE> =
+            td_target.clone() - net.value.forward(net.l1.forward(state.clone()));
+        let delta = delta.array();
+
+        let mut advantage = 0.0;
+        let advantage: Tensor1D<ACTION_SIZE, _> = {
+            let mut data = delta
+                .into_iter()
+                .rev()
+                .map(|delta| {
+                    advantage = GAMMA * LAMBDA * advantage + delta;
+                    advantage
+                })
+                .collect::<Vec<f32>>();
+            data.reverse();
+            dev.tensor(TryInto::<[_; ACTION_SIZE]>::try_into(data.as_slice()).unwrap())
+        };
+        // </calc advantage>
+
+        let (mu, std, value)/* (
+            Tensor1D<ACTION_SIZE, OwnedTape<Cpu>>,
+            Tensor1D<ACTION_SIZE, OwnedTape<Cpu>>,
+            Tensor1D<ACTION_SIZE, OwnedTape<Cpu>>,
+        )*/ = net.forward(state.trace());
+        let log_prob: Tensor1D<ACTION_SIZE, OwnedTape<Cpu>> = {
+            let variance = std.retaped::<OwnedTape<Cpu>>().powi(2);
+            -(action.trace() - mu).powi(2) / (variance * 2.0) - std.ln() - (2.0f32 * f32::consts::PI).sqrt().ln()
+        }/*log_prob(std, mu, action)*/;
+
+        // ratio = P(action | state, pi_net) / P(action | state, target_pi_net)
+        // but compute in log space and then do .exp() to bring it back out of log space
+        let ratio = (log_prob - old_log_prob).exp();
+
+        // because we need to re-use `ratio` a 2nd time, we need to do some tape manipulation here.
+        let surr1: Tensor1D<ACTION_SIZE, OwnedTape<Cpu>> =
+            ratio.with_empty_tape() * advantage.clone();
+        let surr2: Tensor1D<ACTION_SIZE, OwnedTape<Cpu>> =
+            ratio.clamp(1.0 - EPS_CLIP, 1.0 + EPS_CLIP) * advantage.clone();
+
+        let loss = -surr2.minimum(surr1) + smooth_l1_loss(value, td_target, 1.0).array();
+        let loss: Tensor0D<OwnedTape<Cpu>> = loss.mean();
+
+        let loss_v = loss.array();
+
+        // TODO: How to nn.utils.clip_grad_norm_(self.parameters(), 1.0)
+
+        // run backprop
+        let gradients = loss.backward();
+
+        // update weights with optimizer
+        optimizer
+            .update(&mut net, gradients)
+            .expect("Failed to update");
```

Comment:
I assume I'm doing something wrong. Still don't quite understand when and how to handle the tape thing. Correct me if I am wrong but conceptually the tape is like some sort of history so that we can trace the loss backwards all the way to some sort of beginning. How this applies in this example is still not clear though. I have tried to look at examples/rl-ppo and the [python example](https://github.com/seungeunrho/minimalRL/blob/master/ppo-continuous.py) to try to guesstimate the tapes things but still...

```
thread 'main' panicked at 'Failed to update: UnusedParams(UnusedTensors { ids: [UniqueId(0), UniqueId(1), UniqueId(6), UniqueId(7)] })', examples/rl-ppo-continuous.rs:203:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```